README.md
# Bitwarden Alfred Workflow

> Access your Bitwarden passwords, secrets, attachments and more via this powerful Alfred Workflow

# Table of contents

- [Bitwarden Alfred Workflow](#bitwarden-alfred-workflow)
  - [Features](#features)
  - [Installation](#installation)
  - [PATH configuration](#path-configuration)
  - [Usage](#usage)
  - [Login via APIKEY](#login-via-apikey)
  - [Search- / Filtermode](#search---filtermode)
  - [Enable auto background sync](#enable-auto-background-sync)
  - [Enable auto lock](#enable-auto-lock)
  - [Advanced Features / Configuration](#advanced-features--configuration)
  - [Modifier Actions Explained](#modifier-actions-explained)
- [Develop locally](#develop-locally)
- [Licensing and Thanks](#licensing-and-thanks)
  - [Contributors](#contributors)
  - [Source that helped me to get started](#source-that-helped-me-to-get-started)
- [Troubleshooting](#troubleshooting)

## Features

* Completely rewritten in go
* fast secret / item search thanks to caching (no secrets are cached only the keys/names)
  * cache is encrypted
* access to (almost) all object information via this workflow
* download attachments via this workflow
* show favicons of the websites
* auto update
* auto Bitwarden sync in the background
* auto lock on startup and after customizable idle time
* uses the [awgo](https://pkg.go.dev/github.com/deanishe/awgo?tab=doc) framework/library
* many customizations possible


> Requires Alfred 4.1 or newer; NOT tested with Alfred 3

![Bitwarden V2 - Alfred Workflow Demo](./assets/bitwarden-v2.gif)

## Installation
- [Download the latest release](https://github.com/blacs30/bitwarden-alfred-workflow/releases)
- Open the downloaded file in Finder
- Make sure that the [Bitwarden CLI](https://github.com/bitwarden/cli#downloadinstall) version 1.19 or newer is installed
- If running on macOS Catalina or later, you _**MUST**_ add Alfred to the list of security exceptions for running unsigned software. See [this guide](https://github.com/deanishe/awgo/wiki/Catalina) for instructions on how to do this.
  - <sub>Yes, this sucks and is annoying, but there is unfortunately is no easy way around this. macOS requires a paying Developer account for proper app notarization. I'm afraid I'm not willing to pay a yearly subscription fee to Apple just so that this (free and open source) project doesn't pester macOS Gatekeeper.</sub>

## PATH configuration

In many cases the bw and node executables are located in paths outside of the default system PATH.<br>
Please configure the Alfred Worklow variables PATH accordingly.<br>
In a normal terminal type `which bw` and copy the dirname (everything except the "bw") into the PATH workflow variable.<br>
The best is to append it to the existing string and separate it by a colon (:)<br>
Repeat the above steps for node, starting with `which node`.

![Workflow PATH config](./assets/workflow-path-config.gif)

## Usage
To use, activate Alfred and type `.bw` to trigger this workflow. From there:

- type `.bwauth` for login/logout/unlock/lock
- type `.bwconfig` for settings/sync/workflow update/help/issue reports
- type any search term to search for secrets/notes/identities/cards
- modifier keys and actions are presented in the subtitle, different actions are available depending on the object type

## Login via APIKEY
Since version 2.4.1 the workflow supports login via the api key.<br>
Get/create an api key via the web ui. See more information here [https://bitwarden.com/help/article/cli/#using-an-api-key](https://bitwarden.com/help/article/cli/#using-an-api-key)<br>
To use the api key login flow in the workflow set the workflow variable `USE_APIKEY` to true.<br>
The workflow will then ask you for the client_id and client secret to login.<br>
Immediately afterwards it will also ask to unlock with the master password to get a session key. <br>
That is a separate step and required with the api key as login method.

## Search- / Filtermode

Up to version < 2.1.0 the *Fuzzy filtering a la Sublime Text* was default. Starting with version 2.1.0 Alfreds internal filtering is default.

You can change the search-/filtermode yourself easily. This gif shows the 3 steps which need to be done for it:
![Change filter mode](./assets/change-filter-mode.gif)

## Enable auto background sync

In version 2.3.0 the background sync mechanism was added.<br>
It is using the macOS user LaunchAgent.

To install the sync configure the workflow variables:

- `AUTOSYNC_TIMES`, this can be used to configure comma separated multiple sync times per day, e.g. `8:15,23:45`
- alternatively you can use `AUTO_HOUR` together with `AUTO_MIN` for only one sync time

Bitwarden needs to be unlocked for sync to work.

Install via Alfred keyword: `.bwauto`

## Enable auto lock

In version 2.3.0 the background lock and lock on startup mechanism was added.<br>
It is using the macOS user LaunchAgent.

To install the sync configure the workflow variables:

- `LOCK_TIMEOUT` set to a time in minutes after which the workflow should be locked if it hasn't been used in the meantime

The LaunchAgent checks every 5 minutes if the lock timeout has been reached.

The LaunchAgent checks also on load (e.g. startup of the system and login of the user),<br>
if the startup happened within the last 5 minutes, if so then it locks the Bitwarden workflow.

Install via Alfred keyword: `.bwautolock`

## Advanced Features / Configuration

- Configurable [workflow environment variables](https://www.alfredapp.com/help/workflows/advanced/variables/#environment)

| Name                      | Comment                                                                                                                                                                                                                             | Default Value                                                                       |
| ------------------------- | ----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- | ----------------------------------------------------------------------------------- |
| 2FA_ENABLED               | enables or disables 2FA for login (can be set via .bwconfig )                                                                                                                                                                       | true                                                                                |
| 2FA_NODE                  | sets the mode for the 2FA (can be set via .bwconfig ), 0 app, 1, email (not tested), 2 duo (not tested), 3 yubikey (not tested), 4 U2F (not tested)                                                                                 | 0                                                                                   |
| AUTO_HOUR                 | sets the hour for the backround sync to run (is installed separately with .bwauto)                                                                                                                                                  | 10                                                                                  |
| AUTO_MIN                  | sets the minute for the backround sync to run (is installed separately with .bwauto)                                                                                                                                                | 0                                                                                   |
| AUTOSYNC_TIMES            | sets multiple times when bitwarden should sync with the server, this is used first and instead of AUTO_MIN and AUTO_HOUR                                                                                                            | 8:15,23:45                                                                          |
| AUTO_FETCH_ICON_CACHE_AGE | This defines how often the Workflow should check for an icon if is missing, it doesn't need to do it on every run hence this cache                                                                                                  | 1440 (1 day)                                                                        |
| BW_EXEC                   | defines the binary/executable for the Bitwarden CLI command                                                                                                                                                                         | bw                                                                                  |
| BW_DATA_PATH              | sets the path to the Bitwarden Cli data.json                                                                                                                                                                                        | "~/Library/Application Support/Bitwarden CLI/data.json""                            |
| bw_keyword                | defines the keyword which opens the Bitwarden Alfred Workflow                                                                                                                                                                       | .bw                                                                                 |
| bwf_keyword               | defines the keyword which opens the folder search of the Bitwarden Alfred Workflow                                                                                                                                                  | .bwf                                                                                |
| bwauth_keyword            | defines the keyword which opens the Bitwarden authentications of the Alfred Workflow                                                                                                                                                | .bwauth                                                                             |
| bwauto_keyword            | defines the keyword which opens the Bitwarden background sync agent                                                                                                                                                                 | .bwauto                                                                             |
| bwautolock_keyword        | defines the keyword which opens the Bitwarden background lock agent                                                                                                                                                                 | .bwautolock                                                                         |
| bwconf_keyword            | defines the keyword which opens the Bitwarden configuration/settings of the Alfred Workflow                                                                                                                                         | .bwconfig                                                                           |
| DEBUG                     | If enabled print additional debug information, specially about for the decryption process                                                                                                                                           | false                                                                               |
| EMAIL                     | the email which to use for the login via the Bitwarden CLI, will be read from the data.json of the Bitwarden CLI if present                                                                                                         | ""                                                                                  |

CHANGELOG.md
*In compliance with the [GPL-2.0](https://opensource.org/licenses/GPL-2.0) license: I declare that this version of the program contains my modifications, which can be seen through the usual "git" mechanism.*  


2022-09  
Contributor(s):  
Claas Lisowski  
>improve login unlock; open in webui (#137)* check if userid value is nil, set empty string* Move login and unlock into go code* Unlock and sync after api login; remove cache expire checks* add webui_url ui config* fix lint/vet errors  
>Unlock and sync after api login; remove cache expire checks  
>add webui_url config  
>Move login and unlock into go code  
- - - - - - - - - - - - - - - - - - - - - - - - - - - 


2022-03  
Contributor(s):  
blacs30  
Claas Lisowski  
>add option to choose to copy or open url (#125)  
>add instruction to clear cache  
>add option to choose to copy or open url  
>Feature/support v1.21.1 skip types (#123)* skip listed types, support v1.21.1  
>skip listed types, support v1.21.1  
- - - - - - - - - - - - - - - - - - - - - - - - - - - 


2022-02  
Contributor(s):  
Trevor Cohen  
>Update README to specify that path to node executable must be in PATH (#121)If your `bw` and `node` executables are in different directories (e.g.if you use Homebrew to install `bw` and `nvm` for Node) then both thosepaths need to be added to the PATH workflow variable.  
- - - - - - - - - - - - - - - - - - - - - - - - - - - 


2021-12  
Contributor(s):  
blacs30  
Claas Lisowski  
>change default keys back to .bw prefix, don't run auto syncs only show hint  
>use both macos-11 and macos-11.15 for build envs  
>support apikey login (#116)* support apikey login* use both macos-11 and macos-11.15 for build envs* update workflows* update go.* files* update version to 2.4.1* Update Makefile* move go files to src directory* change default keys back to .bw prefix, don't run auto syncs only show hint* Update README.md  
>Update README.md  
>update readme  
- - - - - - - - - - - - - - - - - - - - - - - - - - - 


2021-08  
Contributor(s):  
blacs30  
Claas Lisowski  
>Improve login, move assets (#101)* Update README keywords, remove login and unlock from go code* move assets icons scripts etc directly into the workflow dir  
>Update README keywords, remove login and unlock from go code  
- - - - - - - - - - - - - - - - - - - - - - - - - - - 


2021-06  
Contributor(s):  
blacs30  
Claas Lisowski  
>remove separate cache cli option, merge it into sync. use vars for login and unlock  
>merge master into branch  
>remove separate cache cli option, merge it into sync. use vars for login and unlock (#79)  
>add autolock service, update autosync service  
>add background sync via a macos launchagent  
>Add background sync (#80)* add background sync via a macos launchagent  
>update readme and fix startup time check  
>add Auto lock service (#81)* remove separate cache cli option, merge it into sync. use vars for login and unlock* add background sync via a macos launchagent* update info.plist with vars for auto lock and auto sync* add autolock service, update autosync service* fix go lint* update readme and fix startup time check  
- - - - - - - - - - - - - - - - - - - - - - - - - - - 


2021-02  
Contributor(s):  
Claas Lisowski  
>add 2 troubleshooting topics (#62)* add 2 troubleshooting topics* add steps to update encryption keys  
- - - - - - - - - - - - - - - - - - - - - - - - - - - 


2020-12  
Contributor(s):  
Claas Lisowski  
>Fix/encryption and sync (#58)* Fix encryption process and disable sync per default.  
- - - - - - - - - - - - - - - - - - - - - - - - - - - 


2020-10  
Contributor(s):  
Claas Lisowski  
Luke Hamburg  
>Fix link in README (#48)Fix link - remove extra 'w'  
>Add new feature to decryption secrets from file. (#52)* Add custom path for data.json and move the load function into main/config.go* update cli help text* Add new feature to decrypt secrets from data.json file.* Increment version to 2.2.0  
- - - - - - - - - - - - - - - - - - - - - - - - - - - 


2020-09  
Contributor(s):  
Claas Lisowski  
>Add gif to demonstrate filter mode change. (#44)  
>Feature/custom actions and alfred filter (#43)* Remove dead code* Update readme: changes in filtering and modifier actions.* Remove previous search options.* Update workflow and its version to 2.1.0* Remove space in config.* add: title with user option; modifier actions; folder keyword* README update with thanks to contributors and a note regarding the original fork.* Make the linter happy  
- - - - - - - - - - - - - - - - - - - - - - - - - - - 


2020-08  
Contributor(s):  
Claas Lisowski  
>Add Bitwarden Alfred workflow v2 changes. (#21)  
>Increment version to 2.0.4. Use envconfig now. (#34)  
>Update changes config names in readme. (#37)  
>smaller fixes, demo gif (#24)* Add demo gif to README.* Increase version to 2.0.1 and update workflow description* Add debug print when unexpected cmd error occurs. * Use correct message for locked error.* Use previous bundle id.  
>adding full path to osascript; updating version to 2.0.3 (#30)Add info that for Alfred 4 version 4.1 is requiredrollback to old js for the go internal js prompt - only cli not workflow related  
>Improve config (#39)* Add keywords to config type* Add describtion to readme on how to update the path + gif* Fix totp message; hardcode osascript path; fix issue for login without 2fa* Fix issue for login without 2fa* Update version to 2.0.5  
- - - - - - - - - - - - - - - - - - - - - - - - - - - 


2020-07  
Contributor(s):  
Christian Niehaves  
blacs30  

config.go
// Copyright (c) 2020 Claas Lisowski <github@lisowski-development.com>
// MIT Licence - http://opensource.org/licenses/MIT

package main

import (
	"encoding/json"
	"fmt"
	"github.com/blacs30/bitwarden-alfred-workflow/alfred"
	"github.com/kelseyhightower/envconfig"
	"os"
	"strings"
	"time"

	"log"

	aw "github.com/deanishe/awgo"
)

// Valid modifier keys used to specify alternate actions in Script Filters.
const (
	ModCmd   aw.ModKey = "cmd"   // Alternate action for ⌘↩
	ModAlt   aw.ModKey = "alt"   // Alternate action for ⌥↩
	ModOpt   aw.ModKey = "alt"   // Synonym for ModAlt
	ModCtrl  aw.ModKey = "ctrl"  // Alternate action for ^↩
	ModShift aw.ModKey = "shift" // Alternate action for ⇧↩
	ModFn    aw.ModKey = "fn"    // Alternate action for fn↩
)

var (
	conf      config
	mod1      []aw.ModKey
	mod1Emoji string
	mod2      []aw.ModKey
	mod2Emoji string
	mod3      []aw.ModKey
	mod3Emoji string
	mod4      []aw.ModKey
	mod4Emoji string
	bwData    BwData
)

type modifierActionContent struct {
	Title        string
	Subtitle     string
	Notification string
	Action       string
	Action2      string
	Action3      string
	Arg          string
	Icon         *aw.Icon
	ActionName   string
}

type modifierActionRelation struct {
	ModKey  []aw.ModKey
	Content modifierActionContent
}

type itemActions struct {
	NoMod modifierActionRelation
	Mod1  modifierActionRelation
	Mod2  modifierActionRelation
	Mod3  modifierActionRelation
	Mod4  modifierActionRelation
}

type itemsModifierActionRelation struct {
	Item1 itemActions
	Item2 itemActions
	Item3 itemActions
	Item4 itemActions
	More  modifierActionRelation
}

type config struct {
	// From workflow environment variables
	AutoFetchIconCacheAge    int `default:"1440" split_words:"true"`
	AutoFetchIconMaxCacheAge time.Duration
	BwconfKeyword            string
	BwauthKeyword            string
	BwKeyword                string
	BwfKeyword               string
	BwExec                   string `split_words:"true"`
	// BwDataPath default is set in loadBitwardenJSON()
	BwDataPath         string `envconfig:"BW_DATA_PATH"`
	CacheAge           int    `default:"1440" split_words:"true"`
	Debug              bool   `envconfig:"DEBUG" default:"false"`
	Email              string
	EmptyDetailResults bool `default:"false" split_words:"true"`
	IconCacheAge       int  `default:"43200" split_words:"true"`
	IconCacheEnabled   bool `default:"true" split_words:"true"`
	IconMaxCacheAge    time.Duration
	MaxResults         int `default:"1000" split_words:"true"`
	MaxCacheAge        time.Duration
	Mod1               string `envconfig:"MODIFIER_1" default:"alt"`
	Mod1Action         string `envconfig:"MODIFIER_1_ACTION" default:"username,code"`
	Mod2               string `envconfig:"MODIFIER_2" default:"shift"`
	Mod2Action         string `envconfig:"MODIFIER_2_ACTION" default:"url"`
	Mod3               string `envconfig:"MODIFIER_3" default:"cmd"`
	Mod3Action         string `envconfig:"MODIFIER_3_ACTION" default:"totp"`
	Mod4               string `envconfig:"MODIFIER_4" default:"cmd,alt,ctrl"`
	Mod4Action         string `envconfig:"MODIFIER_4_ACTION" default:"more"`
	NoModAction        string `envconfig:"NO_MODIFIER_ACTION" default:"password,card"`
	OutputFolder       string `default:"" split_words:"true"`
	Path               string
	ReorderingDisabled bool   `default:"true" split_words:"true"`
	Server             string `envconfig:"SERVER_URL" default:"https://bitwarden.com"`
	Sfa                bool   `envconfig:"2FA_ENABLED" default:"true"`
	SfaMode            int    `envconfig:"2FA_MODE" default:"0"`
	SyncCacheAge       int    `default:"1440" split_words:"true"`
	SyncMaxCacheAge    time.Duration
	TitleWithUser      bool `envconfig:"TITLE_WITH_USER" default:"false"`
}

type BwData struct {
	path             string
	InstalledVersion string                 `json:"installedVersion"`
	UserEmail        string                 `json:"userEmail"`
	UserId           string                 `json:"userId"`
	AccessToken      string                 `json:"accessToken"`
	RefreshToken     string                 `json:"refreshToken"`
	ProtectedKey     string                 `json:"__PROTECTED__key"`
	KeyHash          string                 `json:"keyHash"`
	EncKey           string                 `json:"encKey"`
	EncPrivateKey    string                 `json:"encPrivateKey"`
	SecurityStamp    string                 `json:"securityStamp"`
	EncOrgKeys       map[string]interface{} `json:"encOrgKeys"`
	Kdf              int                    `json:"kdf"`
	KdfIterations    int                    `json:"kdfIterations"`
	Unused           map[string]interface{} `json:"-"`
}

func loadBitwardenJSON() error {
	bwDataPath := conf.BwDataPath
	if bwDataPath == "" {
		homedir, err := os.UserHomeDir()
		if err != nil {
			return err
		}
		bwDataPath = fmt.Sprintf("%s/Library/Application Support/Bitwarden CLI/data.json", homedir)
		debugLog(fmt.Sprintf("bwDataPath is: %s", bwDataPath))
	}
	if err := loadDataFile(bwDataPath); err != nil {
		return err
	}
	return nil
}

func loadDataFile(path string) error {
	bwData.path = path
	f, err := os.Open(path)
	if os.IsNotExist(err) {
		return nil
	} else if err != nil {
		return err
	}
	defer f.Close()
	if err := json.NewDecoder(f).Decode(&bwData); err != nil {
		return err
	}
	return nil
}

func loadConfig() {
	// Load workflow vars
	err := envconfig.Process("", &conf)
	if err != nil {
		log.Fatal(err.Error())
	}

	// load the bitwarden data.json
	err = loadBitwardenJSON()
	if err != nil {
		log.Print(err.Error())
	}

	conf.Email = alfred.GetEmail(wf, conf.Email, bwData.UserEmail)
	conf.OutputFolder = alfred.GetOutputFolder(wf, conf.OutputFolder)

	// Set a few cache timeout durations
	setItemCacheAge := conf.CacheAge
	if conf.CacheAge < 30 && conf.CacheAge != 0 {
		setItemCacheAge = 30
	}
	cacheAgeDuration := time.Duration(setItemCacheAge)
	conf.MaxCacheAge = cacheAgeDuration * time.Minute

	iconCacheAgeDuration := time.Duration(conf.IconCacheAge)
	conf.IconMaxCacheAge = iconCacheAgeDuration * time.Minute

	autoFetchIconCacheAgeDuration := time.Duration(conf.AutoFetchIconCacheAge)
	conf.AutoFetchIconMaxCacheAge = autoFetchIconCacheAgeDuration * time.Minute

	// if SYNC_CACHE_AGE is lower than 30 but not 0 set to 30
	setSyncCacheAge := conf.SyncCacheAge
	if conf.SyncCacheAge < 30 && conf.SyncCacheAge != 0 {
		setSyncCacheAge = 30
	}
	syncCacheAgeDuration := time.Duration(setSyncCacheAge)
	conf.SyncMaxCacheAge = syncCacheAgeDuration * time.Minute

	conf.BwauthKeyword = os.Getenv("bwauth_keyword")
	conf.BwconfKeyword = os.Getenv("bwconf_keyword")
	conf.BwKeyword = os.Getenv("bw_keyword")
	conf.BwfKeyword = os.Getenv("bwf_keyword")

	initModifiers()
}

func initModifiers() {
	mod1 = getModifierKey(conf.Mod1)
	mod1Emoji = getModifierEmoji(conf.Mod1)
	mod2 = getModifierKey(conf.Mod2)
	mod2Emoji = getModifierEmoji(conf.Mod2)
	mod3 = getModifierKey(conf.Mod3)
	mod3Emoji = getModifierEmoji(conf.Mod3)
	mod4 = getModifierKey(conf.Mod4)
	mod4Emoji = getModifierEmoji(conf.Mod4)
}

func getModifierKey(keys string) []aw.ModKey {
	items := strings.Split(keys, ",")
	var collectKeys []aw.ModKey
	for _, item := range items {
		switch item {
		case "cmd":
			collectKeys = append(collectKeys, ModCmd)
		case "alt":
			collectKeys = append(collectKeys, ModAlt)
		case "fn":
			collectKeys = append(collectKeys, ModFn)
		case "opt":
			collectKeys = append(collectKeys, ModOpt)
		case "ctrl":
			collectKeys = append(collectKeys, ModCtrl)
		case "shift":
			collectKeys = append(collectKeys, ModShift)
		}
	}
	return collectKeys
}

func getModifierEmoji(keys string) string {
	items := strings.Split(keys, ",")
	var emojiSlice []string
	for _, item := range items {
		switch item {
		case "cmd":
			emojiSlice = append(emojiSlice, "⌘")
		case "alt":
			emojiSlice = append(emojiSlice, "⌥")
		case "fn":
			emojiSlice = append(emojiSlice, "fn")
		case "opt":
			emojiSlice = append(emojiSlice, "⌥")
		case "ctrl":
			emojiSlice = append(emojiSlice, "ˆ")
		case "shift":
			emojiSlice = append(emojiSlice, "⇧")
		}
	}
	emojiString := strings.Join(emojiSlice, "+")

	return emojiString
}

func getTypeEmoji(itemType string) (string, error) {
	modKeysMap := map[string]string{
		conf.NoModAction: "",
		conf.Mod1Action:  mod1Emoji,
		conf.Mod2Action:  mod2Emoji,
		conf.Mod3Action:  mod3Emoji,
		conf.Mod4Action:  mod4Emoji,
	}
	for keys, emoji := range modKeysMap {
		splitKeys := strings.Split(keys, ",")
		for _, key := range splitKeys {
			key = strings.TrimSpace(key)
			if key == itemType {
				return emoji, nil
			}
		}
	}
	return "", fmt.Errorf("no matching key found for type: %s", itemType)
}

func getModifierActionRelations(item Item, itemType string, icon *aw.Icon, totp string, url string) itemsModifierActionRelation {
	var itemModConfig itemsModifierActionRelation
	setModAction(&itemModConfig, item, itemType, "nomod", conf.NoModAction, icon, totp, url)
	setModAction(&itemModConfig, item, itemType, "mod1", conf.Mod1Action, icon, totp, url)
	setModAction(&itemModConfig, item, itemType, "mod2", conf.Mod2Action, icon, totp, url)
	setModAction(&itemModConfig, item, itemType, "mod3", conf.Mod3Action, icon, totp, url)
	setModAction(&itemModConfig, item, itemType, "mod4", conf.Mod4Action, icon, totp, url)
	return itemModConfig
}

func setModAction(itemConfig *itemsModifierActionRelation, item Item, itemType string, modMode string, actionString string, icon *aw.Icon, totp string, url string) {
	// get emojis assigned to the modification key
	moreEmoji, err := getTypeEmoji("more")
	if err != nil {
		log.Fatal(err.Error())
	}
	codeEmoji, err := getTypeEmoji("code")
	if err != nil {
		log.Fatal(err.Error())
	}
	cardEmoji, err := getTypeEmoji("card")
	if err != nil {
		log.Fatal(err.Error())
	}
	passEmoji, err := getTypeEmoji("password")
	if err != nil {
		log.Fatal(err.Error())
	}
	userEmoji, err := getTypeEmoji("username")
	if err != nil {
		log.Fatal(err.Error())
	}
	splitActions := strings.Split(actionString, ",")
	for _, action := range splitActions {
		action = strings.TrimSpace(action)
		if itemType == "item1" {
			title := item.Name
			if conf.TitleWithUser {
				title = fmt.Sprintf("%s - %s", item.Name, item.Login.Username)
			}
			if action == "password" {
				subtitle := "Copy password"
				if modMode == "nomod" {
					subtitle = fmt.Sprintf("↩ or ⇥ copy Password, %s %s, %s %s %s Show more", userEmoji, item.Login.Username, totp, url, moreEmoji)
				}
				modItem := modifierActionContent{
					Title:        title,
					Subtitle:     subtitle,
					Notification: fmt.Sprintf("Copy Password for user:\n%s", item.Login.Username),
					Action:       "-getitem",
					Action2:      fmt.Sprintf("-id %s", item.Id),
					Action3:      " ",
					Arg:          "login.password",
					Icon:         icon,
					ActionName:   action,
				}
				setItemMod(itemConfig, modItem, itemType, modMode)
			}
			if action == "username" {
				assignedIcon := iconUser
				subtitle := "Copy Username"
				if modMode == "nomod" {
					assignedIcon = icon
					subtitle = fmt.Sprintf("↩ or ⇥ copy Username, %s Password, %s %s %s Show more", passEmoji, totp, url, moreEmoji)
				}
				modItem := modifierActionContent{
					Title:        title,
					Subtitle:     subtitle,
					Notification: fmt.Sprintf("Copy Username:\n%s", item.Login.Username),
					Action:       "output",
					Action2:      " ",
					Action3:      " ",
					Arg:          item.Login.Username,
					Icon:         assignedIcon,
					ActionName:   action,
				}
				setItemMod(itemConfig, modItem, itemType, modMode)
			}
			if action == "url" {
				if len(item.Login.Uris) == 0 {
					continue
				}
				assignedIcon := iconLink
				subtitle := "Copy URL"
				if modMode == "nomod" {
					assignedIcon = icon
					subtitle = fmt.Sprintf("↩ or ⇥ copy URL, %s Password, %s Username %s %s Show more", passEmoji, userEmoji, totp, moreEmoji)
				}
				modItem := modifierActionContent{
					Title:        title,
					Subtitle:     subtitle,
					Notification: " ",
					Action:       "-open",
					Action2:      " ",
					Action3:      " ",
					Arg:          item.Login.Uris[0].Uri,
					Icon:         assignedIcon,
					ActionName:   action,
				}
				setItemMod(itemConfig, modItem, itemType, modMode)
			}
			if action == "totp" {
				if totp == "" {
					continue
				}
				assignedIcon := iconUserClock
				subtitle := "Copy TOTP"
				if modMode == "nomod" {
					assignedIcon = icon
					subtitle = fmt.Sprintf("↩ or ⇥ copy TOTP, %s Password, %s Username %s %s Show more", passEmoji, userEmoji, url, moreEmoji)
				}
				modItem := modifierActionContent{
					Title:        title,
					Subtitle:     subtitle,
					Notification: fmt.Sprintf("Copy TOTP for user:\n%s", item.Login.Username),
					Action:       "-getitem",
					Action2:      "-totp",
					Action3:      fmt.Sprintf("-id %s", item.Id),
					Arg:          " ",
					Icon:         assignedIcon,
					ActionName:   action,
				}
				setItemMod(itemConfig, modItem, itemType, modMode)
			}
		}
		if itemType == "item2" {
			modItem := modifierActionContent{
				Title:        item.Name,
				Subtitle:     fmt.Sprintf("↩ or ⇥ copy Note, %s show more", moreEmoji),
				Notification: "Copy Note",
				Action:       "-getitem",
				Action2:      fmt.Sprintf("-id %s", item.Id),
				Action3:      " ",
				Arg:          "notes",
				Icon:         iconNote,
				ActionName:   "",
			}
			setItemMod(itemConfig, modItem, itemType, "nomod")
		}
		if itemType == "item3" {
			title := item.Name
			if conf.TitleWithUser {
				title = fmt.Sprintf("%s - %s", item.Name, item.Card.Number)
			}
			if action == "card" {
				subtitle := "Copy Card Number"
				if modMode == "nomod" {
					subtitle = fmt.Sprintf("%s, %s, ↩ or ⇥ copy Card Number, %s copy Security Code, %s show more", item.Card.Brand, item.Card.Number, codeEmoji, moreEmoji)
				}
				modItem := modifierActionContent{
					Title:        title,
					Subtitle:     subtitle,
					Notification: fmt.Sprintf("Copied Card %s:\n%s", item.Card.Brand, item.Card.Number),
					Action:       "-getitem",
					Action2:      fmt.Sprintf("-id %s", item.Id),
					Action3:      " ",
					Arg:          "card.number",
					Icon:         iconCreditCard,
					ActionName:   action,
				}
				setItemMod(itemConfig, modItem, itemType, modMode)
			}
			if action == "code" {
				subtitle := "Copy card security code"
				if modMode == "nomod" {
					subtitle = fmt.Sprintf("%s, %s, ↩ or ⇥ copy Security Code, %s copy Card Number, %s show more", item.Card.Brand, item.Card.Number, cardEmoji, moreEmoji)
				}
				assignedIcon := iconPassword
				if modMode == "nomod" {
					assignedIcon = iconCreditCard
				}
				modItem := modifierActionContent{
					Title:        title,
					Subtitle:     subtitle,
					Notification: "Copied Card Security Code",
					Action:       "-getitem",
					Action2:      fmt.Sprintf("-id %s", item.Id),
					Action3:      " ",
					Arg:          "card.code",
					Icon:         assignedIcon,
					ActionName:   action,
				}
				setItemMod(itemConfig, modItem, itemType, modMode)
			}
		}
		if itemType == "item4" {
			modItem := modifierActionContent{
				Title:        item.Name,
				Subtitle:     fmt.Sprintf("↩ or ⇥ copy name %s %s, %s show more", item.Identity.FirstName, item.Identity.LastName, moreEmoji),
				Notification: fmt.Sprintf("Copied Identity Name:\n%s %s", item.Identity.FirstName, item.Identity.LastName),
				Action:       "output",
				Action2:      " ",
				Action3:      " ",
				Arg:          " ",
				Icon:         iconIdBatch,
				ActionName:   "",
			}
			setItemMod(itemConfig, modItem, itemType, "nomod")
		}
		if action == "more" {
			modItem := modifierActionContent{
				Title:        item.Name,
				Subtitle:     "Show item",
				Notification: " ",
				Action:       fmt.Sprintf("-id %s", item.Id),
				Action2:      " ",
				Action3:      " ",
				Arg:          " ",
				Icon:         iconList,
				ActionName:   action,
			}
			setItemMod(itemConfig, modItem, itemType, modMode)
		}
	}
}

func setItemMod(itemConfig *itemsModifierActionRelation, content modifierActionContent, itemType string, modMode string) {
	switch itemType {
	case "item1":
		switch modMode {
		case "nomod":
			itemConfig.Item1.NoMod = modifierActionRelation{ModKey: nil, Content: content}
		case "mod1":
			itemConfig.Item1.Mod1 = modifierActionRelation{ModKey: mod1, Content: content}
		case "mod2":
			itemConfig.Item1.Mod2 = modifierActionRelation{ModKey: mod2, Content: content}
		case "mod3":
			itemConfig.Item1.Mod3 = modifierActionRelation{ModKey: mod3, Content: content}
		case "mod4":
			itemConfig.Item1.Mod4 = modifierActionRelation{ModKey: mod4, Content: content}
		}
	case "item2":
		switch modMode {
		case "nomod":
			itemConfig.Item2.NoMod = modifierActionRelation{ModKey: nil, Content: content}
		case "mod1":
			itemConfig.Item2.Mod1 = modifierActionRelation{ModKey: mod1, Content: content}
		case "mod2":
			itemConfig.Item2.Mod2 = modifierActionRelation{ModKey: mod2, Content: content}
		case "mod3":
			itemConfig.Item2.Mod3 = modifierActionRelation{ModKey: mod3, Content: content}
		case "mod4":
			itemConfig.Item2.Mod4 = modifierActionRelation{ModKey: mod4, Content: content}
		}
	case "item3":
		switch modMode {
		case "nomod":
			itemConfig.Item3.NoMod = modifierActionRelation{ModKey: nil, Content: content}
		case "mod1":
			itemConfig.Item3.Mod1 = modifierActionRelation{ModKey: mod1, Content: content}
		case "mod2":
			itemConfig.Item3.Mod2 = modifierActionRelation{ModKey: mod2, Content: content}
		case "mod3":
			itemConfig.Item3.Mod3 = modifierActionRelation{ModKey: mod3, Content: content}
		case "mod4":
			itemConfig.Item3.Mod4 = modifierActionRelation{ModKey: mod4, Content: content}
		}
	case "item4":
		switch modMode {
		case "nomod":
			itemConfig.Item4.NoMod = modifierActionRelation{ModKey: nil, Content: content}
		case "mod1":
			itemConfig.Item4.Mod1 = modifierActionRelation{ModKey: mod1, Content: content}
		case "mod2":
			itemConfig.Item4.Mod2 = modifierActionRelation{ModKey: mod2, Content: content}
		case "mod3":
			itemConfig.Item4.Mod3 = modifierActionRelation{ModKey: mod3, Content: content}
		case "mod4":
			itemConfig.Item4.Mod4 = modifierActionRelation{ModKey: mod4, Content: content}
		}
	}
}
